icu4c/source/common/unicode/platform.h

```diff
@@ -528,6 +528,23 @@ namespace std {
 #   define U_FALLTHROUGH
 #endif
 
+/**
+ * \def UPRV_UNREACHABLE
+ * Annotate unreachable code.
+ * https://clang.llvm.org/docs/LanguageExtensions.html#builtin-unreachable
+ * @internal
+*/
+#if defined(UPRV_UNREACHABLE)
+    // Use the predefined value.
+#elif (defined(__GNUC__) && (U_GCC_MAJOR_MINOR >= 405)) || ( defined(__clang__))
```

Comment:
I am thinking that we might actually want to keep the `U_ASSERT(false)` part if `U_DEBUG` is set to true.
This would allow for the assert to fire and catch anything in Debug builds in case the supposed "unreachable" code is every actually reached.